File: rust/bron_kerbosch/src/core/graph_degeneracy.rs
use super::fortified_counter::FortifiedCounter;
use super::graph::Graph;
use super::priority_queue::Priority;
use super::priority_queue::PriorityQueue;
use super::vertex::{Vertex, VertexMap};
use super::vertexsetlike::VertexSetLike;
use std::iter::FusedIterator;
use std::ops::Not;

// Enumerate connected vertices in degeneracy order, skipping vertices
// whose neighbours have all been enumerated already.
// Along with the vertex picked, includes the subset of neighbours already picked.
pub fn degeneracy_iter<VertexSet>(graph: &Graph<VertexSet>) -> DegeneracyOrderIter<'_, VertexSet>
where
    VertexSet: VertexSetLike,
{
    let mut left_to_pick = FortifiedCounter::new();
    let mut priority_per_vertex = VertexMap::new(None, graph.order());
    let mut queue = PriorityQueue::new(graph.max_degree());
    for v in graph.vertices() {
        if let Some(priority) = Priority::new(graph.degree(v)) {
            left_to_pick.add(v);
            priority_per_vertex[v] = Some(priority);
            queue.put(priority, v);
        }
    }

    DegeneracyOrderIter {
        graph,
        left_to_pick,
        priority_per_vertex,
        queue,
    }
}

pub struct DegeneracyOrderIter<'a, VertexSet>
where
    VertexSet: VertexSetLike,
{
    graph: &'a Graph<VertexSet>,
    left_to_pick: FortifiedCounter<Vertex>,
    priority_per_vertex: VertexMap<Option<Priority>>,
    // If priority is None, the vertex either:
    // - was always irrelevant (unconnected);
    // - was already picked itself;
    // - had all its neighbours picked.
    queue: PriorityQueue<Vertex>,
}

impl<VertexSet> DegeneracyOrderIter<'_, VertexSet>
where
    VertexSet: VertexSetLike,
{
    fn is_consistent(&self) -> bool {
        self.priority_per_vertex
            .iter()
            .all(|(v, &priority)| match priority {
                None => self.left_to_pick.contains(v).not(),
                Some(priority) => self.left_to_pick.contains(v) && self.queue.contains(priority, v),
            })
    }

    fn evaluate_neighbours(&mut self, pick: Vertex) -> VertexSet {
        let mut picked_neighbours = VertexSet::new();
        self.graph.neighbours(pick).for_each(|v| {
            if let Some(old_priority) = self.priority_per_vertex[v] {
                debug_assert!(self.left_to_pick.contains(v));
                debug_assert!(self.queue.contains(old_priority, v));
                let new_priority = Priority::new(old_priority.get() - 1);
                // Because either the new priority takes precedence or the vertex is dequeued,
                // we don't need to find and remove the original entry in the queue.
                // We'll just skip the vertex when the old entry gets popped, and thanks to
                // left_to_pick, we might not even get to popping it at all.
                self.priority_per_vertex[v] = new_priority;
                if let Some(new_priority) = new_priority {
                    self.queue.put(new_priority, v);
                } else {
                    // We discount this neighbour already, but logically it will
                    // be (silently) picked only after we yield the current pick.
                    // So it does not belong in the current picked_neighbours.
                    self.left_to_pick.remove(v);
                }
            } else {
                picked_neighbours.insert(v);
            }
        });
        picked_neighbours
    }
}

impl<VertexSet> FusedIterator for DegeneracyOrderIter<'_, VertexSet> where VertexSet: VertexSetLike {}

impl<VertexSet> Iterator for DegeneracyOrderIter<'_, VertexSet>
where
    VertexSet: VertexSetLike,
{
    type Item = (Vertex, VertexSet);

    fn next(&mut self) -> Option<Self::Item> {
        while self.left_to_pick.count() > 0 {
            debug_assert!(self.is_consistent());
            let pick = self.queue.pop().expect("Cannot pop more than was put");
            if self.priority_per_vertex[pick].take().is_some() {
                self.left_to_pick.remove(pick);
                let picked_neighbours = self.evaluate_neighbours(pick);
                debug_assert!(picked_neighbours.len() < self.graph.degree(pick));
                return Some((pick, picked_neighbours));
            }
        }
        None
    }
}

File: kotlin/src/main/be/steinsomers/bron_kerbosch/GraphDegeneracy.kt
package be.steinsomers.bron_kerbosch

internal data class GraphDegeneracyItem(val pick: Vertex, val pickedNeighbours: MutableSet<Vertex>)

internal class GraphDegeneracy(private val graph: UndirectedGraph) : Iterator<GraphDegeneracyItem> {
    // Possible values of priorityPerVertex (after initialization):
    //   0: never queued because not connected (degree 0),
    //      or no longer queued because it has been yielded itself,
    //      or no longer queued because all neighbours have been yielded
    //   1 or more: candidates queued with priority (degree - #of yielded neighbours)
    private val priorityPerVertex = IntArray(graph.order)
    private val queue = SimplePriorityQueue<Vertex>(graph.maxDegree)
    private var numLeftToPick = 0

    init {
        graph.connectedVertices().forEach { v ->
            val priority = graph.degree(v)
            priorityPerVertex[v.index] = priority
            numLeftToPick += 1
            queue.put(priority, v)
        }
    }

    override fun hasNext(): Boolean {
        return numLeftToPick > 0
    }

    override fun next(): GraphDegeneracyItem {
        while (true) {
            Debug.assert { hasNext() }
            Debug.assert { priorityPerVertex.indices.all { v -> queue.ensure(priorityPerVertex[v], Vertex(v)) } }

            val pick = queue.pop()
            if (priorityPerVertex[pick.index] != 0) {
                priorityPerVertex[pick.index] = 0
                val neighbours = graph.neighbours(pick)
                val pickedNeighbours: MutableSet<Vertex> = HashSet(neighbours.size)
                for (v in neighbours) {
                    val oldPriority = priorityPerVertex[v.index]
                    if (oldPriority != 0) {
                        val newPriority = oldPriority - 1
                        // Requeue with a more urgent priority or dequeue.
                        // Don't bother to remove the original entry from the queue,
                        // since the vertex will be skipped when popped, and thanks to
                        // numLeftToPick we might not need to pop it at all.
                        priorityPerVertex[v.index] = newPriority
                        if (newPriority != 0) {
                            queue.put(newPriority, v)
                        } else {
                            // We discount this neighbour already, but logically it will
                            // be (silently) picked only after we yield the current pick.
                            // So it does not belong in the current pickedNeighbours.
                            numLeftToPick -= 1
                        }
                    } else {
                        pickedNeighbours.add(v)
                    }
                }
                numLeftToPick -= 1
                assert(numLeftToPick >= 0)
                Debug.assert { pickedNeighbours.size < graph.degree(pick) }
                return GraphDegeneracyItem(pick = pick, pickedNeighbours = pickedNeighbours)
            }
        }
    }

    private class SimplePriorityQueue<T>(maxPriority: Int) {
        private val stackPerPriority = Array<ArrayList<T>>(size = maxPriority) { _ -> ArrayList() }

        fun put(priority: Int, elt: T) {
            stackPerPriority[priority - 1].add(elt)
        }

        fun pop(): T {
            for (stack in stackPerPriority) {
                if (stack.isNotEmpty()) {
                    val last = stack.size - 1
                    val elt = stack[last]
                    stack.removeAt(last)
                    return elt
                }
            }
            throw NoSuchElementException("attempt to pop more than was put")
        }

        // Inefficiently check that the queue contains the element at the right priority, if any
        fun ensure(priority: Int, elt: T): Boolean {
            return priority == 0 || stackPerPriority[priority - 1].contains(elt)
        }
    }
}
